server/src/main/java/org/elasticsearch/search/vectors/KnnVectorQueryBuilder.java

```diff
@@ -169,8 +175,15 @@ public KnnVectorQueryBuilder addFilterQueries(List<QueryBuilder> filterQueries)
     @Override
     protected void doWriteTo(StreamOutput out) throws IOException {
         out.writeString(fieldName);
-        out.writeVInt(numCands);
-
+        if (out.getTransportVersion().onOrAfter(TransportVersions.KNN_QUERY_NUMCANDS_AS_OPTIONAL_PARAM)) {
+            out.writeOptionalVInt(numCands);
+        } else {
+            if (numCands == null) {
+                throw new IllegalArgumentException("[" + NUM_CANDS_FIELD.getPreferredName() + "] field cannot be null");
+            } else {
+                out.writeVInt(numCands);
+            }
```

Comment:
Good stuff, we cannot serialize to older nodes and we should just indicate that the value must be provided, by passing any talk of old nodes, etc.

Reply:
Thanks @benwtrent ! Updated the exception message to be a bit more clear on why it's now failing.
```
"[" + NUM_CANDS_FIELD.getPreferredName() + "] field was a mandatory field in previous releases and is required to be non-null by some nodes in this cluster. Please make sure to provide the parameter as part of the request."
```